src/dune_engine/rules.ml

```diff
@@ -171,10 +173,16 @@ module Produce = struct
   end
 end
 
-let produce_dir ~dir rules =
+let of_dir_rules ~dir rules =
   match Dir_rules.Nonempty.create rules with
-  | None -> Memo.Build.return ()
-  | Some rules -> produce (Path.Build.Map.singleton dir rules)
+  | None -> Path.Build.Map.empty
+  | Some rules -> Path.Build.Map.singleton dir rules
+
+let of_rules rules =
+  List.map rules ~f:(fun rule -> (rule.Rule.dir, Dir_rules.Rule rule))
+  |> Path.Build.Map.of_list_multi
+  |> Path.Build.Map.filter_map ~f:(fun rules ->
+         Dir_rules.of_list rules |> Dir_rules.Nonempty.create)
```

Comment:
Seems like this is worth deforesting a little:

```ocaml
val of_list_multi_map : 'a list -> f:('a -> (key * 'v)) -> 'a list t
```

Reply:
Agreed. I still needed two passes with `of_list_multi_map`, so I ended up with a custom traversal